tools/autograd/templates/Functions.cpp

```diff
@@ -1164,6 +1164,97 @@ std::tuple<Tensor, Tensor, Tensor> batchnorm_double_backward(
 
 }
 
+std::tuple<Tensor, Tensor, Tensor, Tensor> bilinear_double_backward(const Tensor& grad_out_grad_input1, const Tensor& grad_out_grad_input2,
```

Comment:
This function is really complicated and will be hard to maintain. Couldn't we stop unrolling backwards at `bilinear_backward` and let autograd figure out the next derivative? Would that be much slower than if we were to use this?

Reply:
So based on Martin's comments on slack, the memory implication of using matrix multiplication are prohibitive.
- So to avoid this and not use inplace operations one easy way could be computing the slices and use `cat` to construct the output.
- A more elaborate way would be to figure out how to deal with the different summations generically. Then you are essentially producing a "reduced einsum" implementation. I'm looking into this for einsum (and would prefer to have exponential time instead of exponential space by using a cuda kernel with an arbitrary number of params).
- the path of least resistance might be keeping it as done here and include a comment to use `einsum` in backward in the future to eliminate the double backward.

For context: The 0.3.1 version that Marcin posted does not do second derivatives, but works for use cases where the version currently in master does not work due to the excessive use of GPU memory and is slow. So if we mainly aim to keep 0.3.1 functionality, we could also drop double differentiation.

I would be glad to follow your advice / choice here.

Best regards

Thomas